library/core/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,74 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else calling the function
+/// via that function pointer is Undefined Behavior. ABI compatibility is a lot stricter than merely
+/// having the same memory layout; for example, even if `i32` and `f32` have the same size and
+/// alignment, they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible call ABI (as
+/// declared via `extern "ABI"`), must take the same number of arguments, and the individual
+/// argument types and the return types must be ABI-compatible.
+///
+/// The call ABIs are guaranteed to be compatible if they are the same, or if the caller ABI is
+/// `$X-unwind` and the callee ABI is `$X`.
+///
+/// The following types are guaranteed to be ABI-compatible:
+///
+/// - `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>` (specifically, only `Box<T, Global>`),
+///   `NonNull<T>` are all ABI-compatible with each other for all `T`.
+/// - `usize` is ABI-compatible with the `uN` integer type of the same size, and likewise `isize` is
+///   ABI-compatible with the `iN` integer type of the same size.
```

Comment:
Do we want to promise that `char` is ABI-compatible with `u32`?

Reply:
I would say so, yes. The docs on `char` talk a lot about its numeric representation. It does have a set of [invalid values](https://doc.rust-lang.org/stable/std/primitive.char.html#validity), but so does `NonZeroI32` for example, and the section at the bottom handles this well.

Reply:
> The docs on char talk a lot about its numeric representation.

They do, but they don't say if it's signed or unsigned -- and for ABI that matters.

Reply:
I'd say let's just skip `char` for now until some requests ABI guarantees for it.